lib/portlayer/storage/nfs/volume.go

```diff
@@ -0,0 +1,260 @@
+		return err
+	}
+
+	//XXX: what should we do if we lose the nfs server between volume and metadata deletion?
+
+	//remove volume metadata directory and children
+	if err := target.RemoveAll(v.volMetadataDirPath(vol.ID)); err != nil {
+		op.Errorf("Failed to remove metadata for volume (%s) at path (%q) on volume store (%s)", vol.ID, v.volDirPath(vol.ID), v.Name)
+		//FIXME: Should we bail here? the volume is gone at this point...
+	}
+	op.Infof("Successfully removed volume (%s) from volumestore (%s)", vol.ID, v.Name)
+
+	return nil
+}
+
+func (v *NFSv3VolumeStore) VolumesList(op trace.Operation) ([]*storage.Volume, error) {
+	volumes := make([]*storage.Volume, 0)
+
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	defer v.MH.Unmount(target)
+
+	volFileInfo, err := target.ReadDir(v.volumesDir())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, volFileInfo := range volFileInfo {
+
+		volMetadata, err := getMetadata(op, v.volMetadataDirPath(volFileInfo.Name()), target)
+		if err != nil {
+			return nil, err //perhaps wait until the end of the loop to report this?
+		}
+
+		volDeviceBacking := NewNFSVolumeDevice(v.Target, v.volDirPath(volFileInfo.Name()))
+
+		vol, err := storage.NewVolume(v.SelfLink, volFileInfo.Name(), volMetadata, volDeviceBacking)
+		if err != nil {
+			return nil, err
+		}
+
+		volumes = append(volumes, vol)
+	}
+
+	return volumes, nil
+}
+
+func writeMetadata(op trace.Operation, metadataPath string, info map[string][]byte, target NFSTarget) error {
+	//NOTE: right now we do not support updating metadata, thus we make the ID directory here
+	_, err := target.MkDir(metadataPath, filePermissions)
+	if err != nil {
+		return err
+	}
+
+	op.Infof("Attempting to write metadata to (%s)", metadataPath)
+	for fileName, data := range info {
+		targetPath := path.Join(metadataPath, fileName)
+		blobFile, err := target.Create(targetPath, filePermissions)
+		// XXX: we might want to make sure the file we want to write does not exist...
+		if err != nil {
+			return err
+		}
+
+		_, err = blobFile.Write(data)
+		if err != nil {
+			//XXX: Should we attempt to clean up what we wrote in this case?
+			return err
+		}
+	}
+	op.Infof("Successfully wrote metadata to (%s)", metadataPath)
+	return nil
+}
+
+func getMetadata(op trace.Operation, metadataPath string, target NFSTarget) (map[string][]byte, error) {
+	op.Infof("Attempting to volume metadata at (%s)", metadataPath)
+	metadataInfo := make(map[string][]byte)
+	dataKeys, err := target.ReadDir(metadataPath)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, metadataFile := range dataKeys {
+		targetPath := path.Join(metadataPath, metadataFile.Name())
+
+		//we need to make sure it is there since we are using open.
+		_, err := target.Lookup(targetPath)
+		if err != nil {
+			return nil, err // FIXME: refine errors later...
+		}
+
+		fileBlob, err := target.Open(targetPath)
```

Comment:
Yes I previously had them below. I think I deleted the calls to close but forgot to add the defer at the time. Thanks for catching that.